homeassistant/components/konnected.py

```diff
@@ -0,0 +1,318 @@
+            }],
+        }),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+DEPENDENCIES = ['http', 'discovery']
+
+ENDPOINT_ROOT = '/api/konnected'
+UPDATE_ENDPOINT = (ENDPOINT_ROOT + r'/device/{device_id:[a-zA-Z0-9]+}')
+
+
+async def async_setup(hass, config):
+    """Set up the Konnected platform."""
+    cfg = config.get(DOMAIN)
+    if cfg is None:
+        cfg = {}
+
+    access_token = cfg.get(CONF_ACCESS_TOKEN)
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {CONF_ACCESS_TOKEN: access_token}
+
+    async def async_device_discovered(service, info):
+        """Call when a Konnected device has been discovered."""
+        _LOGGER.debug("Discovered a new Konnected device: %s", info)
+        host = info.get(CONF_HOST)
+        port = info.get(CONF_PORT)
+
+        device = KonnectedDevice(hass, host, port, cfg)
+        await device.async_setup()
+
+    discovery.async_listen(
+        hass,
+        SERVICE_KONNECTED,
+        async_device_discovered)
+
+    hass.http.register_view(KonnectedView(access_token))
+
+    return True
+
+
+class KonnectedDevice(object):
+    """A representation of a single Konnected device."""
+
+    def __init__(self, hass, host, port, config):
+        """Initialize the Konnected device."""
+        self.hass = hass
+        self.host = host
+        self.port = port
+        self.user_config = config
+
+        import konnected
+        self.client = konnected.Client(host, str(port))
+        self.status = self.client.get_status()
+        _LOGGER.info('Initialized Konnected device %s', self.device_id)
+
+    async def async_setup(self):
+        """Set up a newly discovered Konnected device."""
+        user_config = self.config()
+        if user_config:
+            _LOGGER.debug('Configuring Konnected device %s', self.device_id)
+            self.save_data()
+            await self.async_sync_device_config()
+            await discovery.async_load_platform(
+                self.hass, 'binary_sensor',
+                DOMAIN, {'device_id': self.device_id})
+            await discovery.async_load_platform(
+                self.hass, 'switch', DOMAIN,
+                {'device_id': self.device_id})
+
+    @property
+    def device_id(self):
+        """Device id is the MAC address as string with punctuation removed."""
+        return self.status['mac'].replace(':', '')
+
+    def config(self):
+        """Return an object representing the user defined configuration."""
+        device_id = self.device_id
+        valid_keys = [device_id, device_id.upper(),
+                      device_id[6:], device_id.upper()[6:]]
+        configured_devices = self.user_config[CONF_DEVICES]
+        return next((device for device in
+                     configured_devices if device[CONF_ID] in valid_keys),
+                    None)
+
+    def save_data(self):
+        """Save the device configuration to `hass.data`.
+
+        TODO: This can probably be refactored and tidied up.
```

Comment:
Since this is a new component we should make sure to have it as ready as possible. At the least remove todo comment, at best follow the todo 👍